JAX: Refactor JAX D2L API to support BatchNorm (#2354)
d2l-ai/d2l-en: chapter_linear-classification/classification.md

## The `Classifier` Class

> :begin_tab:`pytorch, mxnet, tensorflow`

Reviewer: ```suggestion
```

> We define the `Classifier` class below. In the `validation_step` we report both the loss value and the classification accuracy on a validation batch. We draw an update for every `num_val_batches` batches. This has the benefit of generating the averaged loss and accuracy on the whole validation data. These average numbers are not exactly correct if the last batch contains fewer examples, but we ignore this minor difference to keep the code simple.

Reviewer: ```suggestion
```

Author: This was done intentionally; if I do not add the explicit tabs for other frameworks, the rendering will be off.

> :end_tab:

Reviewer: ```suggestion
```

> :begin_tab:`jax`

Reviewer: ```suggestion
```

We define the `Classifier` class below. In the `validation_step` we report both the loss value and the classification accuracy on a validation batch. We draw an update for every `num_val_batches` batches. This has the benefit of generating the averaged loss and accuracy on the whole validation data. These average numbers are not exactly correct if the last batch contains fewer examples, but we ignore this minor difference to keep the code simple.

> We also redefine the `training_step` for JAX since all models that will

Reviewer: ```suggestion
:begin_tab:`jax`
We also redefine the `training_step` method for JAX since all models that will
```

> subclass `Classifier` later will have a loss that returns aux data.

Reviewer: ```suggestion
subclass `Classifier` later will have a loss that returns auxiliary data.
```

> This aux data can be used for BatchNorm models explained in :numref:`sec_batch_norm`,

Reviewer: ```suggestion
This auxiliary data can be used for models with batch normalization (to be explained in :numref:`sec_batch_norm`),
```

> in all other cases we'll make the loss also return a placeholder empty dictionary to

Reviewer: ```suggestion
while in all other cases we'll make the loss also return a placeholder (empty dictionary) to
```

> represent the aux data.

Reviewer: ```suggestion
represent the auxiliary data.
```

```{.python .input}
%%tab jax
class Classifier(d2l.Module):  #@save
    def training_step(self, params, batch, state):
        # Here value is a tuple since models with BatchNorm layers require

>         # the loss to return aux data.

Reviewer: ```suggestion
        # the loss to return auxiliary data
```

        value, grads = jax.value_and_grad(
            self.loss, has_aux=True)(params, *batch[:-1], batch[-1], state)
        l, _ = value
        self.plot("loss", l, train=True)
        return value, grads

    def validation_step(self, params, batch, state):

>         # Discard the second return value. Used for trainig models with

Reviewer: ```suggestion
        # Discard the second returned value. It is used for training models with
```

>         # a BatchNorm layer, since loss also returns aux data

Reviewer: ```suggestion
        # BatchNorm layers since loss also returns auxiliary data
```
